names.lua

```diff
@@ -1,55 +1,117 @@
+namescr = defclass(namescr, gui.Screen)
+namescr.focus_path = 'names'
+function namescr:init()
+	if not dfhack.gui.getViewscreenByType(df.viewscreen_layer_choose_language_namest)==true then
+		self:addviews{
+			widgets.Label{
+				view_id='namescr',
+				frame = {b=4, l=1},
+				text = {
+					{text = "Press f to Change First Name"},NEWLINE,
+					{text = "Press Esc to Set Name and Exit"},
+						},
+					}
+				}
+		vw = dfhack.gui.getCurViewscreen()
+		if df.viewscreen_itemst:is_instance(vw.parent) then
+			fact = vw.parent.item.general_refs[0].artifact_id
+			trg = df.artifact_record.find(fact)
+		elseif df.viewscreen_dungeon_monsterstatusst:is_instance(vw.parent) then
+			uid = vw.parent.unit.id
+			trg = df.unit.find(uid)
+		elseif df.global.ui_advmode.menu==1 then
+			local t_look=df.global.ui_look_list.items[df.global.ui_look_cursor]
+			if t_look.type==2 then
+				trg=t_look.unit
+			end
+		end
+		choices = df.viewscreen_setupadventurest:new()
+		choices.page = 7
+		local tn = choices.adventurer
+		utils.assign(tn.name, trg.name)
+		gui.simulateInput(choices, 'A_CUST_NAME')
+		vw = dfhack.gui.getViewscreenByType(df.viewscreen_layer_choose_language_namest)
+	elseif dfhack.gui.getViewscreenByType(df.viewscreen_layer_choose_language_namest)==true then
+		qerror('names screen already displayed')
+	end
+	return vw,trg
 end
-
-if args.item then
-	fact = dfhack.gui.getCurViewscreen().parent.item.general_refs[0].artifact_id
-	trg = df.artifact_record.find(fact)
+function namescr:setName()
+	for k = 0,6 do
+		trg.name.words[k] = vw.name.words[k]
+		trg.name.parts_of_speech[k] = vw.name.parts_of_speech[k]
+		trg.name.language = vw.name.language
+		trg .name.has_name = vw.name.has_name
+	end
 end
- 
-if args.unit then
-	trg = dfhack.gui.getCurViewscreen().parent.unit
+function namescr:setFirst()
+local str = ''
+	dlg.showInputPrompt("Set First Name?","First: ",COLOR_WHITE,'',
+		function(str)
+			if str==nil then
+				self:callback("setFirst")
+			else
+				vw.name.first_name = str
+				trg.name.first_name = str
+			end
+		end)
+	return str
 end
-
-if args.first then
-	trg.name.first_name = args.first
+function namescr:onRenderBody(dc)
+    self._native.parent:render()
 end
-
-function newName()
- 	newn = dfhack.gui.getCurViewscreen().name
-	oldn = trg.name
-	for k = 0,6 do
-	 	oldn.words[k] = newn.words[k]
- 		oldn.parts_of_speech[k] = newn.parts_of_speech[k]
+function namescr:onInput(keys)
+	if keys.SELECT then
+		self:setName()
+	end
+	if keys.CUSTOM_F then
+		self:setFirst()
+		if not str==nil then
+			trg.name.first_name = str
+		end
+	end
+	if keys.LEAVESCREEN then
+		self:setName()
+		self:dismiss()
+		dfhack.screen.dismiss(self._native.parent)
 	end
- 	oldn.language = newn.language
- 	oldn.has_name = newn.has_name
+	return
+	gui.simulateInput(self._native.parent, keys)
```

Comment:
I just learned about that one recently too. I was mostly just confused because I couldn't see where input was being sent to the parent at all, because I thought the "return" statement didn't return anything.